### PR TITLE
Kvs benchmark

### PR DESCRIPTION
**Pull request by zzlk**
# caveats
- Doesn't seem to work with multithreaded tokio runtime, bug in implementation of tcp operators?
- CPU usage of server caps out at around 90%, client at 70%. I should be able to increase QD until one of them caps out, but why not? Logic bug in client code, maybe it's not actually achieving the stated QD.

# benchmark stuff

Server was run with:
```
cargo flamegraph -c "record --call-graph dwarf -F 4000 -g" --example kvs_bench --profile=profile -- --role=server --addr=127.0.0.1:40000
```

client was run with:
```
cargo run --example kvs_bench --release -- --role=client --addr=127.0.0.1:40000
```

flamegraph was generated from server. On my machine this produced ~320k gets/s

Unfortunately github doesn't display the svg file in a very useful way, download it and open it with chrome so you can click and drill down on specific items.

[flamegraph download](https://drive.google.com/file/d/1tcjVY6JdBCSp0dmfrC1k-Lnmzbw8Ab7B/view)

flamegraph

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
https://gist.github.com/MingweiSamuel/63f51359483175517be88bf7d77930fe

File: hydroflow/examples/kvs_bench/client.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+pub async fn run_client(server_addr: SocketAddr) {
+    println!("tid client: {}", palaver::thread::gettid());
+
+    let stream = TcpStream::connect(server_addr).await.unwrap();
+    stream.set_nodelay(true).unwrap();
+
+    println!("connected");
+
+    let (mut outbound, mut inbound) = tcp_bytes(stream);
```
Comment (MingweiSamuel) [inline hydroflow/examples/kvs_bench/client.rs:21]: So one thing, I'm not sure if the Anna KVS benchmark even exercised the networking stack >.>
May be an unfair comparison that way

File: hydroflow/examples/kvs_bench/client.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        if let Some(Ok(response)) = inbound.next().await {
+            let response: KVSResponse = deserialize_from_bytes(response);
+            match response {
+                KVSResponse::Response { key, value } => assert_eq!(value, map[&key]),
+            }
+            outstanding -= 1;
+            gets += 1;
+        }
```
Comment (MingweiSamuel) [inline hydroflow/examples/kvs_bench/client.rs:71]: Ought to do the sends and receives simultaneously? I.e. with a loop and [`select!`](https://docs.rs/tokio/latest/tokio/macro.select.html)

File: hydroflow/examples/kvs_bench/server.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                _ => None
+            }
+        });
+
+        parsed_gets = my_demux[gets] -> filter_map(|m| {
+            match m {
+                KVSRequest::Get{key} => Some((key, ())),
+                _ => None
+            }
+        });
```
Comment (MingweiSamuel) [inline hydroflow/examples/kvs_bench/server.rs:39]: Can simplify with

```rust
            -> demux(|m, var_args!(puts, gets)| match m {
                KVSRequest::Put { key, value } => puts.give((key, value)),
                KVSRequest::Get { key } => gets.give((key, ())),
            });
```
Then no need for the `filter_map`s

File: hydroflow_lang/src/graph/ops/dest_sink_serde_tcp.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        let write_prologue = quote_spanned! {op_span=>
+            let (#send_ident, #recv_ident) = #root::tokio::sync::mpsc::unbounded_channel();
+            df
+                .spawn_task(async move {
+                    use #root::futures::sink::SinkExt;
+
+                    let mut recv = #recv_ident;
+                    let mut sink = #sink_arg;
+                    while let Some(payload) = recv.recv().await {
+                        let item = #root::util::serialize_to_bytes(payload);
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/dest_sink_serde_tcp.rs:54]: Is there a reason you didn't go with?
```rust
-> map(|payload| serialize_to_bytes(payload)) -> dest_sink(outbound)
```

File: hydroflow_lang/src/graph/ops/source_stream_serde_tcp.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                 _| {
+        let receiver = &arguments[0];
+        let stream_ident = wc.make_ident("stream");
+        let write_prologue = quote_spanned! {op_span=>
+            let mut #stream_ident = Box::pin(#receiver);
+        };
+        let write_iterator = quote_spanned! {op_span=>
+            let #ident = std::iter::from_fn(|| {
+                match #root::futures::stream::Stream::poll_next(#stream_ident.as_mut(), &mut std::task::Context::from_waker(&context.waker())) {
+                    std::task::Poll::Ready(Some(std::result::Result::Ok(payload))) => Some(#root::util::deserialize_from_bytes(payload)),
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/source_stream_serde_tcp.rs:50]: Same as above, why not use an extra `map` or `filter_map`?

**Comment by MingweiSamuel** [comment]
Wonder if using a `flatten` iter would work better than the `pop_buffer` buffering

**Comment by zzlk** [comment]
> https://gist.githubusercontent.com/MingweiSamuel/63f51359483175517be88bf7d77930fe/raw/c4d7a17a724c93a296162bd9af57add3445dd6e0/kvs_flamegraph.svg

This one also seems to be broken for me, chrome on linux.

File: hydroflow/examples/kvs_bench/server.rs (hunk 3 above)
Comment (zzlk) [reply hydroflow/examples/kvs_bench/server.rs:39]: Made the change.

File: hydroflow_lang/src/graph/ops/dest_sink_serde_tcp.rs (hunk 4 above)
Comment (zzlk) [reply hydroflow_lang/src/graph/ops/dest_sink_serde_tcp.rs:54]: Lack of experience navigating hydroflow :)

File: hydroflow_lang/src/graph/ops/source_stream_serde_tcp.rs (hunk 5 above)
Comment (zzlk) [reply hydroflow_lang/src/graph/ops/source_stream_serde_tcp.rs:50]: same as above :)

**Comment by MingweiSamuel** [comment]
I might ought to take another look at how to DRY up the `source_stream_serde` and `dest_sink_serde` operators too

**Pull request closed without merging**